### PR TITLE
krita: 5.1.5 -> 5.2.0

### PR DESCRIPTION
**Pull request by nek0**
## Description of changes

This Pull request bumps the version of `krita` to the newest stable version (5.2.0).
Additionally the following dependencies have been initialized in the following version:

- `zug` at 0.1.0
- `lager` at 0.1.0
- `kseexpr`at 4.0.4.0

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/krita/default.nix [hunk 1]
```diff
@@ -1,7 +1,7 @@
 { callPackage, ... } @ args:
 
 callPackage ./generic.nix (args // {
-  version = "5.1.5";
+  version = "5.2.0";
   kde-channel = "stable";
-  sha256 = "1lx4x4affkbh47b7w5qvahkkr4db0vcw6h24nykak6gpy2z5wxqw";
+  sha256 = "15wjia5509rqg2h6kh5a6cd90mdjr7ij4xfcp586ghsri9rijsnk";
```
Comment (eclairevoyant) [inline pkgs/applications/graphics/krita/default.nix:6]: please use `hash`

File: pkgs/by-name/ks/kseexpr/package.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+}:
+
+
+stdenv.mkDerivation rec {
+  pname = "kseexpr";
+  version = "4.0.4.0";
+  src = fetchFromGitLab {
+    domain = "invent.kde.org";
+    owner = "graphics";
+    repo = pname;
```
Comment (eclairevoyant) [inline pkgs/by-name/ks/kseexpr/package.nix:17]:
```suggestion
    repo = "kseexpr";
```
semantically different

File: pkgs/by-name/la/lager/package.nix [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  fetchFromGitHub,
+  boost, cmake, cereal, gcc7, immer, ncurses, pkg-config, sass, SDL2, SDL2_ttf, zug
+}:
+
+stdenv.mkDerivation rec {
+  pname = "lager";
+  version = "0.1.0";
+  src = fetchFromGitHub {
+    owner = "arximboldi";
+    repo = pname;
```
Comment (eclairevoyant) [inline pkgs/by-name/la/lager/package.nix:11]:
```suggestion
    repo = "lager";
```

File: pkgs/by-name/zu/zug/package.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  cmake, boost
+}:
+
+
+stdenv.mkDerivation rec {
+  pname = "zug";
+  version = "0.1.0";
+  src = fetchFromGitHub {
+    owner = "arximboldi";
+    repo = pname;
```
Comment (eclairevoyant) [inline pkgs/by-name/zu/zug/package.nix:12]:
```suggestion
    repo = "zug";
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
also relevant: #246247, #247777, fixes #261012

File: pkgs/by-name/la/lager/package.nix [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  propagatedBuildInputs = [
+    boost
+    cereal
+    immer
+    zug
+  ];
+  meta = with lib; {
+    homepage    = "https://github.com/arximboldi/lager";
+    description = "library for functional interactive c++ programs";
+    license     = licenses.lgpl3Plus;
```
Comment (ramirez7) [inline pkgs/by-name/la/lager/package.nix:39]: This is the Nix from the `lager` repo, but the license they have there is wrong (the library is MIT)

File: pkgs/by-name/zu/zug/package.nix [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  ];
+  cmakeFlags = [
+    "-Dzug_BUILD_TESTS=OFF"
+    "-Dzug_BUILD_EXAMPLES=OFF"
+  ];
+  meta = with lib; {
+    homepage    = "https://github.com/arximboldi/zug";
+    description = "library for functional interactive c++ programs";
+    maintainers = with maintainers; [ nek0 ];
+    license     = licenses.lgpl3Plus;
```
Comment (sifmelcara) [inline pkgs/by-name/zu/zug/package.nix:28]: This should be `boost` license (their .nix file has wrong license)

File: pkgs/by-name/zu/zug/package.nix [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    hash = "sha256-7xTMDhPIx1I1PiYNanGUsK8pdrWuemMWM7BW+NQs2BQ=";
+  };
+  buildInputs = [
+    cmake
+    boost
+  ];
+  cmakeFlags = [
+    "-Dzug_BUILD_TESTS=OFF"
+    "-Dzug_BUILD_EXAMPLES=OFF"
+  ];
```
Comment (sifmelcara) [inline pkgs/by-name/zu/zug/package.nix:23]: No need to disable tests, IIRC they builds

File: pkgs/by-name/la/lager/package.nix [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  cmakeFlags = [
+    "-Dlager_BUILD_TESTS=OFF"
+    "-Dlager_BUILD_EXAMPLES=OFF"
+  ];
+  propagatedBuildInputs = [
+    boost
+    cereal
+    immer
+    zug
+  ];
```
Comment (sifmelcara) [inline pkgs/by-name/la/lager/package.nix:35]: Only
```
  nativeBuildInputs = [ cmake ];
  buildInputs = [ boost immer zug ];
```
Is required for build inputs

File: pkgs/by-name/la/lager/package.nix (hunk 5 above)
Comment (sifmelcara) [reply pkgs/by-name/la/lager/package.nix:39]: cc @arximboldi

File: pkgs/applications/graphics/krita/generic.nix [hunk 9]
```diff
@@ -60,6 +56,8 @@ mkDerivation rec {
   cmakeFlags = [
     "-DPYQT5_SIP_DIR=${python3Packages.pyqt5}/${python3Packages.python.sitePackages}/PyQt5/bindings"
     "-DPYQT_SIP_DIR_OVERRIDE=${python3Packages.pyqt5}/${python3Packages.python.sitePackages}/PyQt5/bindings"
+    "-DBUILD_KRITA_QT_DESIGNER_PLUGINS=ON"
+    "-DBUILD_TESTING=OFF"
```
Comment (sifmelcara) [inline pkgs/applications/graphics/krita/generic.nix:60]: I think it is important to make sure the tests keep passing, especially with all the non-standard build setup in nixpkgs. Any reason why we disable the test here?

**Review by nek0 — COMMENTED** [review]
Thanks for all the comments!

File: pkgs/applications/graphics/krita/default.nix (hunk 1 above)
Comment (nek0) [reply pkgs/applications/graphics/krita/default.nix]: done.

File: pkgs/applications/graphics/krita/generic.nix (hunk 9 above)
Comment (nek0) [reply pkgs/applications/graphics/krita/generic.nix]: Build went through ok on my builder.

File: pkgs/by-name/la/lager/package.nix (hunk 8 above)
Comment (nek0) [reply pkgs/by-name/la/lager/package.nix]: I changed the build inputs accordingly.

File: pkgs/by-name/zu/zug/package.nix (hunk 7 above)
Comment (nek0) [reply pkgs/by-name/zu/zug/package.nix]: test enabled again and build confirmed.

**Review by sifmelcara — APPROVED** [review]

**Comment by sifmelcara** [comment]
Commit order seems wrong, `krita: 5.1.5 -> 5.2.0` should be after `zug`/`lager`/`kseexpr`

File: pkgs/by-name/la/lager/package.nix (hunk 5 above)
Comment (arximboldi) [reply pkgs/by-name/la/lager/package.nix]: Correct, the library is actually MIT, sorry for that mistake.

File: pkgs/by-name/zu/zug/package.nix (hunk 6 above)
Comment (arximboldi) [reply pkgs/by-name/zu/zug/package.nix]: You're right.

**Comment by eclairevoyant** [comment]
Builds fine, and runs fine after some basic testing.

I could nitpick about formatting, but don't really want to, once commits are reordered I'd approve.

**Comment by nek0** [comment]
> Builds fine, and runs fine after some basic testing.
> 
> I could nitpick about formatting, but don't really want to, once commits are reordered I'd approve.

Commits have been reordered. Version update is now after library inits.

**Comment by ramirez7** [comment]
It built for me too! I couldn't run it, but I think that's due to my system:

```
Cannot mix incompatible Qt library (5.15.9) with this library (5.15.10)
Aborted (core dumped)
```

I remember this error but couldn't remember if there's some workaround :thinking:

File: pkgs/by-name/la/lager/package.nix [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  buildInputs = [
+    boost
+    immer
+    zug
+  ];
+  nativeBuildInputs = [
+    cmake
+  ];
+  cmakeFlags = [
+    "-Dlager_BUILD_TESTS=OFF"
```
Comment (delroth) [inline pkgs/by-name/la/lager/package.nix:25]: Is there a specific reason to not build and run the tests? Usually we try to whenever possible, if it's not possible it should be documented with a comment somewhere in this derivation.

File: pkgs/by-name/la/lager/package.nix [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    "-Dlager_BUILD_EXAMPLES=OFF"
+  ];
+  meta = with lib; {
+    homepage    = "https://github.com/arximboldi/lager";
+    description = "library for functional interactive c++ programs";
```
Comment (delroth) [inline pkgs/by-name/la/lager/package.nix:30]: Is this description correct? It's the same as `zug` and doesn't really match what e.g. the GitHub project description says.

(The description for `zug` also doesn't really match its GitHub project description, fwiw - both might need reviewing.)

File: pkgs/by-name/zu/zug/package.nix [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  pname = "zug";
+  version = "0.1.0";
+  src = fetchFromGitHub {
+    owner = "arximboldi";
+    repo = "zug";
+    rev = "v${version}";
+    hash = "sha256-7xTMDhPIx1I1PiYNanGUsK8pdrWuemMWM7BW+NQs2BQ=";
+  };
+  buildInputs = [
+    cmake
```
Comment (delroth) [inline pkgs/by-name/zu/zug/package.nix:17]: cmake should likely be a nativeBuildInputs

File: pkgs/development/libraries/fftw/default.nix [hunk 13]
```diff
@@ -25,6 +25,10 @@ stdenv.mkDerivation (finalAttrs: {
     sha256 = "sha256-VskyVJhSzdz6/as4ILAgDHdCZ1vpIXnlnmIVs0DiZGc=";
   };
 
+  patches = [
+    ./remove_missing_FFTW3LibraryDepends.patch
```
Comment (delroth) [inline pkgs/development/libraries/fftw/default.nix:29]: There should likely be comments in here documenting:
1. Where the patch comes from (backported from upstream? custom?)
2. Why we need it

Depending on (2), this should possibly also be sent upstream. We usually much prefer backports of upstream accepted patches to custom patches that need to be maintained in nixpkgs long term.

File: pkgs/applications/graphics/krita/generic.nix [hunk 14]
```diff
@@ -18,27 +19,22 @@ mkDerivation rec {
...
+    inherit hash;
   };
 
   patches = [
-    (fetchpatch {
-      name = "exiv2-0.28.patch";
-      url = "https://gitlab.archlinux.org/archlinux/packaging/packages/krita/-/raw/acd9a818660e86b14a66fceac295c2bab318c671/exiv2-0.28.patch";
-      hash = "sha256-iD2pyid513ThJVeotUlVDrwYANofnEiZmWINNUm/saw=";
-    })
+    ./find_openjpeg.patch
```
Comment (delroth) [inline pkgs/applications/graphics/krita/generic.nix:26]: Similarly to my comment re: the fftw patch - why do we need this, and can we get it upstreamed?

**Review by delroth — CHANGES_REQUESTED** [review]
1. Could you please clean up the branch's commit history? e.g. make sure that the fixup commits are properly folded into the commit that they are fixing.
2. The amount of rebuilds means this needs to go through staging. Can you re-target this PR? That's probably due to the fftw patch.

File: pkgs/development/libraries/fftw/default.nix (hunk 13 above)
Comment (tobim) [reply pkgs/development/libraries/fftw/default.nix:29]: I a modified form of the patch upstream: https://github.com/FFTW/fftw3/pull/338
You can use `fetchpatch` to get it from there.

Comment (nek0) [reply pkgs/development/libraries/fftw/default.nix:29]: thanks!

**Comment by nek0** [comment]
> 1. Could you please clean up the branch's commit history? e.g. make sure that the fixup commits are properly folded into the commit that they are fixing.
> 
>     2. The amount of rebuilds means this needs to go through staging. Can you re-target this PR? That's probably due to the fftw patch.

I cleaned the commit history as best I could, but I don't know how and where to retarget this PR.

**Review by tobim — CHANGES_REQUESTED** [review]
Please clean up the argument and input lists as suggested.

The easy way to retarget a PR to staging is to close this one and create a completely new PR.

File: pkgs/by-name/la/lager/package.nix (hunk 10 above)
Comment (nek0) [reply pkgs/by-name/la/lager/package.nix]: Tests enabled.

File: pkgs/by-name/zu/zug/package.nix (hunk 12 above)
Comment (nek0) [reply pkgs/by-name/zu/zug/package.nix:20]: cmake has been moved.

File: pkgs/by-name/la/lager/package.nix (hunk 11 above)
Comment (nek0) [reply pkgs/by-name/la/lager/package.nix]: description has been updated.

**Comment by aidalgol** [comment]
Krita 5.2.1 released since this PR was opened: https://krita.org/en/item/krita-5-2-1-released/

**Comment by nek0** [comment]
With the help of @SuperSandro2000, this Pull request now builds `krita` with all plugins.

**Comment by nek0** [comment]
@ofborg build krita

**Review by delroth — APPROVED** [review]
Code change looking good now, thank you so much for the cleanups / fixes!

Running nixpkgs-review right now since ofborg shows the build as broken for x86_64 - I suspect it's staging brokenness and not related to your changes, but just making sure. If I forget to come back to this PR within the next 24h please feel free to ping me again directly :-)

**Comment by delroth** [comment]
(If you can't reproduce this, make sure the patch file isn't cached in your local Nix store - otherwise it won't try to re-fetch it.)

**Review by delroth — CHANGES_REQUESTED** [review]
(As per above.)

**Comment by nek0** [comment]
the commit on `fftw`'s side seems to have changed. I fixed it and it should work now.

**Review by delroth — APPROVED** [review]
`krita` built fine with this PR rebased on master.

**Pull request merged**